src/Collapse.js

```diff
@@ -125,78 +139,86 @@ const defaultProps = {
-    elem.style[dimension] = this._getScrollDimensionValue(elem, dimension);
-  };
-
-  handleEntered = (elem) => {
-    elem.style[this.getDimension()] = null;
-  };
-
-  /* -- Collapsing -- */
-  handleExit = (elem) => {
-    const dimension = this.getDimension();
-    elem.style[dimension] = `${this.props.getDimensionValue(
-      dimension,
-      elem,
-    )}px`;
-    triggerBrowserReflow(elem);
-  };
-
-  handleExiting = (elem) => {
-    elem.style[this.getDimension()] = null;
-  };
-
-  // for testing
-  _getScrollDimensionValue(elem, dimension) {
-    const scroll = `scroll${dimension[0].toUpperCase()}${dimension.slice(1)}`;
-    return `${elem[scroll]}px`;
-  }
-
-  render() {
-    const {
+const Collapse = React.forwardRef(
+  (
+    {
       onEnter,
       onEntering,
       onEntered,
       onExit,
       onExiting,
       className,
       children,
+      dimension,
+      getDimensionValue,
       ...props
-    } = this.props;
+    },
+    ref,
+  ) => {
+    /* -- Expanding -- */
+    const handleEnterAction = useCallback(
+      (elem) => {
+        elem.style[getDimension(dimension)] = '0';
+      },
+      [dimension],
+    );
+
+    const handleEnteringAction = useCallback(
+      (elem) => {
+        const computedDimension = getDimension(dimension);
+        elem.style[dimension] = collapseHelpers._getScrollDimensionValue(
+          elem,
+          computedDimension,
+        );
+      },
+      [dimension],
+    );
+
+    const handleEnteredAction = useCallback(
+      (elem) => {
+        elem.style[getDimension(dimension)] = null;
+      },
+      [dimension],
+    );
+
+    /* -- Collapsing -- */
+    const handleExitAction = useCallback(
+      (elem) => {
+        const computedDimension = getDimension(dimension);
+        elem.style[dimension] = `${getDimensionValue(
+          computedDimension,
+          elem,
+        )}px`;
+        triggerBrowserReflow(elem);
+      },
+      [dimension, getDimensionValue],
+    );
+
+    const handleExitingAction = useCallback(
+      (elem) => {
+        elem.style[getDimension(dimension)] = null;
+      },
+      [dimension],
+    );
 
     delete props.dimension;
     delete props.getDimensionValue;
-
-    const handleEnter = createChainedFunction(this.handleEnter, onEnter);
+    const handleEnter = createChainedFunction(handleEnterAction, onEnter);
     const handleEntering = createChainedFunction(
-      this.handleEntering,
+      handleEnteringAction,
       onEntering,
     );
-    const handleEntered = createChainedFunction(this.handleEntered, onEntered);
-    const handleExit = createChainedFunction(this.handleExit, onExit);
-    const handleExiting = createChainedFunction(this.handleExiting, onExiting);
+    const handleEntered = createChainedFunction(handleEnteredAction, onEntered);
+    const handleExit = createChainedFunction(handleExitAction, onExit);
+    const handleExiting = createChainedFunction(handleExitingAction, onExiting);
```

Comment:
**suggestion:** there's not much point to `useCallback`-ing the wrapped callbacks while not wrapping `createChainedFunction` in anything memoized.

instead of doing this, then, i would wrap the `createChainedFunction` calls in `useMemo`, and not further memoize the individual handlers.

Reply:
Yes, you're right ! I'm fixing it thank you !